src/Scroll/Scroll.man
.\" t
.\" @(#)Scroll.1	02/07/00
.TH Scroll 1.8 "Feb 7 2000" Scroll
.UC
.SH NAME
\fBScroll\fP \- the AfterStep scroll-bar module

.SH SYNOPSIS
\fBScroll\fP [\-f \fIconfig_file\fP] [\-v | \--version] [\-h | \--help]
[x y]
.sp
Although \fBScroll\fP can be run from the command line, it is normally 
spawned by afterstep from within an afterstep menu file by 'Scroll':
.nf
.sp
% cat "~/GNUstep/Library/AfterStep/start/Modules/Scroll"
	Module "Scroll" Scroll
	MiniPixmap "mini-as.xpm"
.fi

.SH DESCRIPTION
The Scroll module prompts the user to select a target window, if
the module was not launched from within a window context in AfterStep.
After that, it adds scroll bars to the selected window, to reduce the
total desktop space consumed by the window.

Scroll should not be used with windows which move or resize themselves, nor
should it be used with windows which set the WM_COLORMAP_WINDOWS property.
Operation is fine with windows that have a private colormap.

.SH OPTIONS
.IP "\-f \fIconfig_file\fP"
Use \fIconfig_file\fP instead of the default configuration file.

.IP "\-h | \--help"
Print a usage message and exit successfully.

.IP "\-v | \--version"
Print version information and exit successfully.

.SH INITIALIZATION
During initialization, \fIScroll\fP will search for a configuration 
file which describes the colors to use.  This file will be the one 
specified by the \fI\-f\fP option (if given).  Else, the 
~/GNUstep/Library/AfterStep/scroll file will be used if it exists.  
Else, /usr/local/share/afterstep/scroll will be used.

If the \fBScroll\fP executable is linked to another name, ie 'ln -s
Scroll OtherScroll, then another module called OtherScroll can be
started, with a completely different configuration than Scroll,
simply by changing the keyword \fI*Scroll\fP to \fI*MoreScroll\fP.  This way
multiple Scrolling modules can be used.

.SH INVOCATION
\fBScroll\fP can be invoked from a menu item as shown in the Synopsis
section above, or from binding "Module Scroll [\fIx y\fP]" to a key-stroke
in a \fIfeel\fP file, where the parameter x and y are integers, which
describe the horizontal and vertical window size reduction.

AfterStep will search through the directories specified in the
\fIModulePath\fP configuration in the \fIbase.#bpp\fP (where # is your
colordepth) to attempt to locate Scroll.

.SH CONFIGURATION OPTIONS
\fBScroll\fP reads the \fIscroll\fP file when it starts up and looks for
it's configuration options:

.IP "*ScrollFore \fIcolor\fP"
Tells the module to use \fIcolor\fP instead of grey for scroll bars 
themselves.

.IP "*ScrollBack \fIcolor\fP"
Tells the module to use \fIcolor\fP instead of black for the window
background.

.SH BUGS
When the scroll bars are removed by clicking on the button in the
lower right corner, the window does not restore its location
correctly.

.SH AUTHOR
Robert Nation
.nf
David Mihm <dmihm@swbell.net> [Man page]

.SH SEE ALSO
.BR afterstep (1)


ChangeLog

2003-09-25  sasha

	* generating ChangeLog

2003-09-24  sasha

	* Completed look.Glass ; renamed PointerFore, PointerBack to
	  CursorFore, CursorBack to match old Cursor option; Moved Cursor
	  definitions from feel into look Completed Colorscheme's support
	  for Cursor's color

2003-09-23  sasha

	* Added look options for mouse pointer coloring: PointerFore and
	  PointerBack; Repaired code to update window's decorations when
	  stickiness flag is toggled;

2003-09-22  sasha

	* Added cursors to default feel file; enable cursor switching while
	  hovering over different elements of the interface; Added code to
	  change color of the cursor according to what colorscheme is set
	  to - may need to tweak a bit , or make colors configurable
	* Fixed locale parsing from LANG variable; Fixed module startup to
	  pass on current locae specified on command line

2003-09-19  sasha

	* added debugging code for cursor loading

2003-09-16  sasha

	* fixed bug with windows teleporting back to original position, if
	  moved/resized too fast

2003-09-15  sasha

	* updated configure scripts
	* Fixed configure.in to properly enable i18n in libAfterImage as
	  requested. Internationalized text should be displayed properly
	  now, even if its not in UTF8; Update src/test to work with new
	  libs and added test code for i18n testing
	* Fixed unquoted string in i18N code reported by Alex Mezin

2003-09-14  sasha

	* Added checks for locale and proper initialization of
	  libAfterImage t3ext drawing in accordance to its value; Fixed
	  problem with EdgeScroll being mangled when just look changes;
	  Fixed recently introduced bug causing wierd corners in framned
	  windows upon the look change
	* Fixed typoe for msg type in SendString being declared as some
	  wierd type of unsigned long long  - possibly causing wierdness on
	  64 bit machines

2003-09-13  sasha

	* implemented feedback functionality via sending the e-mail to
	  sasha :)

2003-09-12  sasha

	* added configure option to send e-mail back to developer on
	  installation; Fixed canvas border width handling while resizing
	  and in Pager;

2003-09-11  sasha

	* fixed handling of WM_PROTOCOL changes where WM_COMMAND would get
	  lost and session management would stop working; Found bug with
	  huge impact - all the geometry calculations were off as we were
	  forgetting border width of the frame window; Fixed handling of
	  button presses on border of the frame

2003-09-10  sasha

	* fixed NeXTish colorscheme somewhat
	* Fixed workspace state save/restore to allow terms to exec command
	  using -e, and to have argv's with spaces in them; Fixed frame
	  corner sizing to make it same width as is tbar size;

2003-09-09  sasha

	* More fixing Pager's geometry; Added -i cmd line parameter to
	  ascompose to allow for 'include' files (such as colorschemes,
	  etc.); Updated configure scripts to properly detect headers when
	  its run from libAfterBase; Updated version to beta2; Updated TEAM
	  file

2003-09-08  sasha

	* Fixed Pager to properly account for label size while sizing
	  min-desktop windows

2003-09-07  sasha

	* somewhat debugged Pager's geometry handling - need to add some
	  code to properly recalculate Pager's geometry if its soecified as
	  ratio of the main desktop vs. exact geometry

2003-09-03  sasha

	* added new AfterStep log icon; fixed wharf config file

2003-09-02  sasha

	* added new xml tag : <color> to define symbolic color names
	* changed handling of tiled images to clip it properly when
	  Right/Left align is used

2003-08-30  sasha

	* debugged workspace state handling code, and improved Wait loop

2003-08-29  sasha

	* mostly implemented new workspace state saving and restoring

2003-08-28  sasha

	* fixed workspace status generator to not save xrm params, as noone
	  seems to support thoise
	* added caching of pixmaps for unfocused bas in window decorations
	  to speed up fast focus switching
	* 64-bit fixes: CARD32 instead of Window type
	* 64-bit fixes: properties and strlen warnings
	* 64-bit fixes: converted intermodule communications to always use
	  32bit values
	* more 64 bit fixes

2003-08-27  sasha

	* 64-bit compilation fixes
	* Fixed slowness in intermodule communications manifesting itself
	  when desktops are switched
	* fixed handling of non-default MyBackgrounds, and proper adding of
	  DeskConfigs while switching desks
	* synching
	*  fixed segfaulting in Pager on look switches or after window
	  restacking
	* Fixed handling of hints flags - it was doing nothing before if
	  flags == 0 - was screwing MOtif Hints like totatlly; Fixed shaped
	  window handling in WinList; Added check to not do too much
	  redrawing when WinLIst is receiving window list - may want to do
	  the same in Pager

2003-08-26  sasha

	*  added parameters to import funcions, to allow for sized
	  image/preview stuff; Removed compression buffer from ASImage
	  structure to save up some memory; Hopefully fixed hints handling
	  on 64 bit hw

2003-08-24  sasha

	* fixed event handling when windows is getting mapped - no more
	  lost events due to not selected mask ; Fixed initial status when
	  mouse pointer is out of the screen; Fixed ICCCM size hints
	  handling to have BaseSize and MinSize interchange each other when
	  one is ommited - fixes GIMPs initial placement

2003-08-23  sasha

	* added new 3 icons to complete the set of default icons: switched
	  AfterStep to use ~/.afterstep instead of ridiculous ~/GNUstel/L/A
	  stuff

2003-08-22  sasha

	* Fixed handling of shaped windows - shape is not erroneously
	  cleared, and client's shape is subtracted from decoration's shape
	* fixed autodetection of zipped xpms and added 6 more icons to
	  default iconset

2003-08-21  sasha

	* more improvements to asimbrowser

2003-08-20  sasha

	* fixed makefile for ascp, when fltk is not found; added better
	  Style for BitchX to database; remnoved scrollbar from BitchX in
	  Wharf

src/Scroll/Scroll.c
/*
 * Copyright (c) 1997 Guylhem Aznar <guylhem@oeil.qc.ca>
 * Copyright (c) 1994 Nobutaka Suzuki <nobuta-s@is.aist-nara.ac.jp>
 * Copyright (c) 1994 Robert Nation
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

#define YES "Yes"
#define NO  "No"

#include "../../configure.h"

#include <stdlib.h>

#include <X11/Xlib.h>
#include <X11/Xutil.h>
#include <X11/cursorfont.h>
#include <X11/Xmu/WinUtil.h>

#define IN_MODULE
#define MODULE_X_INTERFACE
#include "../../include/aftersteplib.h"
#include "../../include/module.h"
#include "Scroll.h"

int fd[2];

ScreenInfo Scr;

char *BackColor = "black";
char *ForeColor = "grey";

Window app_win;

#define MW_EVENTS   (ExposureMask | ButtonReleaseMask | KeyReleaseMask)

void ParseOptions (const char *);

void
version (void)
{
  printf ("%s version %s\n", MyName, VERSION);
  exit (0);
}

void
usage (void)
{
  printf ("Usage:\n"
	  "%s [-v|--version] [-h|--help] [--window window-id] [x y]\n", MyName);
  exit (0);
}

/***********************************************************************
 *
 *  Procedure:
 *	main - start of module
 *
 ***********************************************************************/
int
main (int argc, char **argv)
{
  char configfile[255];
  char *realconfigfile;
  char *temp;
  int i;
  char *global_config_file = NULL;

  /* Save our program name - for error messages */
  temp = strrchr (argv[0], '/');
  MyName = temp ? temp + 1 : argv[0];
  set_application_name( argv[0] );

  for (i = 1; i < argc && *argv[i] == '-'; i++)
    {
      if (!strcmp (argv[i], "-h") || !strcmp (argv[i], "--help"))
	usage ();
      else if (!strcmp (argv[i], "-v") || !strcmp (argv[i], "--version"))
	version ();
      else if (!strcmp (argv[i], "-w") || !strcmp (argv[i], "--window"))
	app_win = strtol (argv[++i], NULL, 16);
      else if (!strcmp (argv[i], "-c") || !strcmp (argv[i], "--context"))
	i++;
      else if (!strcmp (argv[i], "-f") && i + 1 < argc)
	global_config_file = argv[++i];
    }

  /* Dead pipe == dead AfterStep */
  signal (SIGPIPE, DeadPipe);
  set_signal_handler (SIGSEGV);

  x_fd = ConnectX( &Scr, display_name, 0 );
  /* connect to AfterStep */
  fd[0] = fd[1] = ConnectAfterStep (0);

  if (i < argc)
    {
      char *cptr = argv[i++];
      extern int Reduction_H, Reduction_V;
      Reduction_H = strtol (cptr, &cptr, 10);
      Reduction_V = strtol (cptr, &cptr, 10);
      if (Reduction_H < 2)
	Reduction_H = 2;
      if (Reduction_V < 2)
	Reduction_V = 2;
    }

  /* scan config file for set-up parameters */
  /* Colors and fonts */
  if (global_config_file != NULL)
    ParseOptions (global_config_file);
  else
    {
      sprintf (configfile, "%s/scroll", AFTER_DIR);

      realconfigfile = PutHome (configfile);

      if ((CheckFile (realconfigfile)) == -1)
	{
	  free (realconfigfile);
	  sprintf (configfile, "%s/scroll", AFTER_SHAREDIR);
	  realconfigfile = PutHome (configfile);
	}
      ParseOptions (realconfigfile);
      free (realconfigfile);
    }

  if (app_win == 0)
    GetTargetWindow (&app_win);

  if (app_win == 0)
    return 1;

  fd_width = GetFdWidth ();

  GrabWindow (app_win);
  Loop (app_win);

  return 0;
}


void
ParseOptions (const char *filename)
{
  FILE *file;
  char *line, *tline;
  int len;

  if ((file = fopen (filename, "r")) != (FILE *) NULL)
    {
      line = (char *) safemalloc (MAXLINELENGTH);
      len = strlen (MyName);
      while ((tline = fgets (line, MAXLINELENGTH, file)) == NULL)
	{
	  while (isspace (*tline))
	    tline++;
	  if ((*tline == '*') && (!mystrncasecmp (tline + 1, MyName, len)))
	    {
	      tline += len + 1;
	      if (!mystrncasecmp (tline, "Fore", 4))
        ForeColor = stripcpy( tline + 4);
	      else if (!mystrncasecmp (tline, "Back", 4))
        BackColor = stripcpy( tline + 4);
	    }
	}
      free (line);
      fclose (file);
    }
}

/***********************************************************************
 *
 * Detected a broken pipe - time to exit
 *
 **********************************************************************/
void
DeadPipe (int nonsense)
{
  extern Atom wm_del_win;

  XReparentWindow (dpy, app_win, Scr.Root, 0, 0);
  send_wm_protocol_request(app_win, wm_del_win, CurrentTime);
  XSync (dpy, 0);
  exit (0);
}


/**********************************************************************
 *
 * If no application window was indicated on the command line, prompt
 * the user to select one
 *
 *********************************************************************/
void
GetTargetWindow (Window * app_win)
{
  XEvent eventp;
  int val = -10, trials;
  Window target_win;

  trials = 0;
  while ((trials < 100) && (val != GrabSuccess))
    {
      val = XGrabPointer (dpy, Scr.Root, True,
			  ButtonReleaseMask,
			  GrabModeAsync, GrabModeAsync, Root,
			  XCreateFontCursor (dpy, XC_crosshair),
			  CurrentTime);
      if (val != GrabSuccess)
	{
	  sleep_a_little (1000);
	}
      trials++;
    }
  if (val != GrabSuccess)
    {
      fprintf (stderr, "%s: Couldn't grab the cursor!\n", MyName);
      exit (1);
    }
  XMaskEvent (dpy, ButtonReleaseMask, &eventp);
  XUngrabPointer (dpy, CurrentTime);
  XSync (dpy, 0);
  *app_win = eventp.xany.window;
  if (eventp.xbutton.subwindow != None)
    *app_win = eventp.xbutton.subwindow;

  target_win = ClientWindow (*app_win);
  if (target_win != None)
    *app_win = target_win;
}


/****************************************************************************
 *
 * Find the actual application
 *
 ***************************************************************************/
Window
ClientWindow (Window input)
{
  Atom _XA_WM_STATE;
  unsigned int nchildren;
  Window root, parent, *children, target;
  unsigned long nitems, bytesafter;
  unsigned char *prop;
  Atom atype;
  int aformat;
  int i;

  _XA_WM_STATE = XInternAtom (dpy, "WM_STATE", False);

  if (XGetWindowProperty (dpy, input, _XA_WM_STATE, 0L,
			  3L, False, _XA_WM_STATE, &atype,
			  &aformat, &nitems, &bytesafter,
			  &prop) == Success)
    {
      if (prop != NULL)
	{
	  XFree (prop);
	  return input;
	}
    }

  if (!XQueryTree (dpy, input, &root, &parent, &children, &nchildren))
    return None;

  for (i = 0; i < nchildren; i++)
    {
      target = ClientWindow (children[i]);
      if (target != None)
	{
	  XFree ((char *) children);
	  return target;
	}
    }
  XFree ((char *) children);
  return None;
}
